pandas/core/indexes/interval.py

```diff
@@ -1284,21 +1285,40 @@ def equals(self, other):
                 self.right.equals(other.right) and
                 self.closed == other.closed)
 
-    def _setop(op_name):
+    def _setop(op_name, check_subtypes=False):
         def func(self, other):
             msg = ('can only do set operations between two IntervalIndex '
                    'objects that are closed on the same side')
             other = self._as_like_interval_index(other, msg)
+
+            if check_subtypes:
```

Comment:
why wouldn't you always check this?

Reply:
Operations between incompatible subtypes are technically valid for `difference` and `intersection`, since they always return a subset of the original index, so the left dtype can always be maintained:
```python
In [2]: ii1 = pd.interval_range(0, 4)

In [3]: ii2 = pd.interval_range(pd.Timestamp('20180101'), periods=4)

In [4]: ii1.difference(ii2)
Out[4]:
IntervalIndex([(0, 1], (1, 2], (2, 3], (3, 4]]
              closed='right',
              dtype='interval[int64]')

In [5]: ii1.intersection(ii2)
Out[5]:
IntervalIndex([]
              closed='right',
              dtype='interval[int64]')
```
While technically valid, not sure that `difference` and `intersection` really make sense for incompatible subtypes.  Didn't want to immediately jump to raising for a technically valid op, but can certainly do the check for all set ops and always raise for incompatible subtypes, if that's the desired behavior.

Reply:
yes I would rather not have a special check here. we simply don't want to allow ops between invalid types

Reply:
done